### PR TITLE
House Cleaning - Part 2 (Collision and Pathfinding)

### PR DESCRIPTION
**Pull request by lzardy**
Rather than simply cleaning, I decided to also remodel a bit, hopefully this doesn't cause too much of a hassle for the review process.
* Adds comments to NavigationGrid, NavigationGridCell, Projectile, AttackableUnit, BaseTurret, ObjAIBase, CollisionHandler.
* Cleaned unused functions and variables of NavigationGrid.
* Edited ActorList of NavigationGridCell to be an actual list.
* Added ToCenterScaledCoordinates function to MovementVector which is used for PacketNotifier.
* Implemented server-side pathfinding, refer to HandleMove.
* Fixed names of StatModifier functions.
 * Fixed UltimateQuadTree and replaced the NuGet Package with the fixed version.
* Fully implemented CollisionHandler.
* Collision affected ObjAIBase GameObjects will repath upon collision.
* Collision affected GameObjects will teleport out of terrain upon collision (unless it's an ObjAIBase that is dashing).
* Moved TeleportTo from ObjAIBase to GameObject.
* Added the rest of the crowd control functions of ObjAIBase to its interface.
* Edited MovementRequest to reflect LeaguePackets' version.
* Added functions to Extensions: IsVectorValid, GetTriangleArea (laser) DistanceSquaredToRectangle, IsPointInRectangle, and GetCircleEscapePoint.
* Added Turret based functions to ObjectManager for managing BaseTurrets.
* Slightly optimized CheckForTargets function in BaseTurret.
* Overrode GetMoveSpeed for BaseTurret to be sure it never has movementspeed.
* Added OnAdded function to BaseTurret for ObjectManager.
* Added LaneID to BaseTurret along with SetLaneID, used to identify the location of specific turrets.
* Cleaned AddMinionTarget of ApiFunctionManager.
* Added functions: GetCellIndex, GetClosestValidCellIndex, and GetCellsWithinRange to NavigationGrid which are used for getting a better terrain exit (and hopefully pathfinding in the future).
* Added SetOpen function to NavigationGridCell to be used for dynamic pathfinding around cells occupied by actors.
* Removed unnecessary usage of SetTeam by integrating teams into GameObject initialization.
* Integrated OnCollision overrides for other projectile types.
* Particles no longer depend on the Target class.
* Projectiles no longer depend on the Target class.
* Overrode GameObject.Move() for Projectile and refactored to account for target changes.
* Separated ClassifyUnit and NavigationGridCellFlags into GameServerCore.Enums.
* LaneMinions get assigned their models properly.

Solves #1033, #1031, #953, #938, #843, #754, #554, and #65.

Progress #972, #721, #662, #566, #555, #545, and #527.

### CODE REVIEW

File: GameServerLib/GameObjects/Missiles/Projectile.cs
Comment (danil179) [inline GameServerLib/GameObjects/Missiles/Projectile.cs:154]: Isn't this function should be the same as moving with `GameObject`?

File: GameServerLib/GameObjects/Missiles/Projectile.cs
Comment (danil179) [inline GameServerLib/GameObjects/Missiles/Projectile.cs]: I think you can use here `var next = GetTargetPosition()` and remove the next if

File: GameServerLib/GameServerLib.csproj
Comment (danil179) [inline GameServerLib/GameServerLib.csproj]: Do you know where one can find this version? I couldn't find it

File: GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs
Comment (danil179) [inline GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs:701]: Probably add here `TODO: resolve dashes into object/wall`

**Review by danil179 — COMMENTED** [review]
Seems OK apart from some comments. There are too much changes so I'm sure that some of the mechanisms changed and may not work/cause a crash, but I think that this change adds very much to the readability and the changes are where the are too much "special cases". I think this will hep to the incremental development.

I have to check this (especially the navgrid where you removed the use of short which is one of the main conversions. Also some changes in the movement system of particles/missles and the collision of certain objects). After I'll check it I'll be able to approve.

File: GameServerLib/GameObjects/Missiles/Projectile.cs
Comment (lzardy) [reply GameServerLib/GameObjects/Missiles/Projectile.cs:154]: No, because projectiles don't have waypoints, they only have a target or a destination. This new Move function removes the unneeded complexity of waypoints.

File: GameServerLib/GameServerLib.csproj
Comment (lzardy) [reply GameServerLib/GameServerLib.csproj]: https://www.nuget.org/packages/UltimateQuadTree-fix/1.0.5

File: GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs
Comment (lzardy) [reply GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs:701]: Collisions while an ObjAiBase.IsDashing are ignored in ObjAiBase.OnCollision.

Comment (lzardy) [reply GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs:701]: Or did you mean that I should comment "we need an event which is called on collision while dashing"?

File: GameServerLib/GameObjects/Missiles/Projectile.cs
Comment (lzardy) [reply GameServerLib/GameObjects/Missiles/Projectile.cs]: Will be added next commit.
Resolved.

File: GameServerLib/GameServerLib.csproj
Comment (lzardy) [reply GameServerLib/GameServerLib.csproj]: Will be reverted to the original and will update CollisionObject to have their origin at the top left corner of their bounds.

**Review by danil179 — COMMENTED** [review]
Seems that the QuadTree should work now. I'll run this and check for crashes/problems and so on.

**Comment by danil179** [comment]
#1033 not resolved.

**Comment by danil179** [comment]
Collision seems to work fine. Minions completely ignore other minions and do not attack at all. ~~Turrets continue to target even after going out of range.~~

**Comment by lzardy** [comment]
LaneMinion AI is for another PR. Turrets do not continue to target when something leaves their attack range. 
What you may be noticing is actually a desync between client to server. The client is expecting the lane minions to stop at the first turret and start attacking, so client-side, they stop moving, but server-side they are still moving towards the nexus.

File: GameServerLib/GameObjects/Other/CollisionHandler.cs
Comment (danil179) [inline GameServerLib/GameObjects/Other/CollisionHandler.cs]: I think that this change actually sort of remove the purpose of QuadTree. This number is the maximum number of objects on one "partition" of the QuadTree and not the maximum objects in general as far as I know. If there is a problem in the collision system it is maybe somewhere else (probably it is related to the upside down y axis in general in some of the conversions).

**Review by danil179 — COMMENTED** [review]
If you see that something is wrong with the collision system, I don't think it is the max objects variable, but it is probably somewhere else (in the y axis). Maybe we can just generate sort of "visuals" for the bounding boxes and will see if that works.

File: GameServerLib/GameObjects/Other/CollisionHandler.cs
Comment (danil179) [inline GameServerLib/GameObjects/Other/CollisionHandler.cs]: There will be a problem here because the starting point shouldn't be (0,0) but instead some negative value of the NavGrid.

**Review by danil179 — COMMENTED** [review]
Currently, it seems that there are problem with the following things:
1. Healthbar of not LaneMinions and its update
2. Things with vision seems to me not so correct and I don't think it is all due to client desync, but some are "vision related" things. I surmise that the server is able to check all the collisions and turret attacks but the vision is somehow affecting what we see in the client.

I approve it till this point, but I'll merge only after these crucial issues will be resolved (and I don't think that client desync is related to all of them).

**Review by danil179 — COMMENTED** [review]
The code seems correct with no special issues. I'll run and test the game to see if the issues with the minions have been fixed.

**Review by danil179 — APPROVED** [review]
Approved. Tested it and it works fine. The code is much better now. There are some changes in the minions AI but it will improve over time.

**Pull request merged**
